# McGranahanLab/mhc_hammer: Citations

## [nf-core](https://pubmed.ncbi.nlm.nih.gov/32055031/)

> Ewels PA, Peltzer A, Fillinger S, Patel H, Alneberg J, Wilm A, Garcia MU, Di Tommaso P, Nahnsen S. The nf-core framework for community-curated bioinformatics pipelines. Nat Biotechnol. 2020 Mar;38(3):276-278. doi: 10.1038/s41587-020-0439-x. PubMed PMID: 32055031.

## [Nextflow](https://pubmed.ncbi.nlm.nih.gov/28398311/)

> Di Tommaso P, Chatzou M, Floden EW, Barja PP, Palumbo E, Notredame C. Nextflow enables reproducible computational workflows. Nat Biotechnol. 2017 Apr 11;35(4):316-319. doi: 10.1038/nbt.3820. PubMed PMID: 28398311.

## [IMGT/HLA](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC29780/)

> Robinson J, Waller MJ, Parham P, Bodmer JG, Marsh SG. IMGT/HLA Database--a sequence database for the human major histocompatibility complex. Nucleic Acids Res. 2001 Jan 1;29(1):210-3. doi: 10.1093/nar/29.1.210. PMID: 11125094; PMCID: PMC29780.

## Pipeline tools

* [BCFTools](https://pubmed.ncbi.nlm.nih.gov/21903627/)
    > Li H. A statistical framework for SNP calling, mutation discovery, association mapping and population genetical parameter estimation from sequencing data. Bioinformatics. 2011 Nov 1;27(21):2987-93. doi: 10.1093/bioinformatics/btr509. Epub 2011 Sep 8. PMID: 21903627; PMCID: PMC3198575.

* [BEDTools](https://pubmed.ncbi.nlm.nih.gov/20110278/)
    > Quinlan AR, Hall IM. BEDTools: a flexible suite of utilities for comparing genomic features. Bioinformatics. 2010 Mar 15;26(6):841-2. doi: 10.1093/bioinformatics/btq033. Epub 2010 Jan 28. PubMed PMID: 20110278; PubMed Central PMCID: PMC2832824.

* [Bowtie2](https://pubmed.ncbi.nlm.nih.gov/22388286/)
    > Langmead B, Salzberg SL. Fast gapped-read alignment with Bowtie 2. Nat Methods. 2012 Mar 4;9(4):357-9. doi: 10.1038/nmeth.1923. PMID: 22388286; PMCID: PMC3322381.

* [ensembl-VEP](https://pubmed.ncbi.nlm.nih.gov/27268795/)
    > McLaren W, Gil L, Hunt SE, Riat HS, Ritchie GR, Thormann A, Flicek P, Cunningham F. The Ensembl Variant Effect Predictor. Genome Biol. 2016 Jun 6;17(1):122. doi: 10.1186/s13059-016-0974-4. PMID: 27268795; PMCID: PMC4893825.

* [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)

* [GATK](https://pubmed.ncbi.nlm.nih.gov/20644199/)
    > McKenna A, Hanna M, Banks E, Sivachenko A, Cibulskis K, Kernytsky A, Garimella K, Altshuler D, Gabriel S, Daly M, DePristo MA. The Genome Analysis Toolkit: a MapReduce framework for analyzing next-generation DNA sequencing data. Genome Res. 2010 Sep;20(9):1297-303. doi: 10.1101/gr.107524.110. Epub 2010 Jul 19. PMID: 20644199; PMCID: PMC2928508.

* [HLA-HD](https://pubmed.ncbi.nlm.nih.gov/28419628/)
    > Kawaguchi S, Higasa K, Shimizu M, Yamada R, Matsuda F. HLA-HD: An accurate HLA typing algorithm for next-generation sequencing data. Hum Mutat. 2017 Jul;38(7):788-797. doi: 10.1002/humu.23230. Epub 2017 May 12. PMID: 28419628.

* [Jellyfish](https://pubmed.ncbi.nlm.nih.gov/21217122/)
    > Marçais G, Kingsford C. A fast, lock-free approach for efficient parallel counting of occurrences of k-mers. Bioinformatics. 2011 Mar 15;27(6):764-70. doi: 10.1093/bioinformatics/btr011. Epub 2011 Jan 7. PMID: 21217122; PMCID: PMC3051319.

* [Mosdepth](https://pubmed.ncbi.nlm.nih.gov/29096012/)
    > Pedersen BS, Quinlan AR. Mosdepth: quick coverage calculation for genomes and exomes. Bioinformatics. 2018 Mar 1;34(5):867-868. doi: 10.1093/bioinformatics/btx699. PMID: 29096012; PMCID: PMC6030888.

* [NovoAlign](https://www.novocraft.com/products/novoalign/)

* [picard-tools](http://broadinstitute.github.io/picard)

* [SAMtools](https://pubmed.ncbi.nlm.nih.gov/19505943/)
    > Li H, Handsaker B, Wysoker A, Fennell T, Ruan J, Homer N, Marth G, Abecasis G, Durbin R; 1000 Genome Project Data Processing Subgroup. The Sequence Alignment/Map format and SAMtools. Bioinformatics. 2009 Aug 15;25(16):2078-9. doi: 10.1093/bioinformatics/btp352. Epub 2009 Jun 8. PMID: 19505943; PMCID: PMC2723002.

* [STAR](https://pubmed.ncbi.nlm.nih.gov/23104886/)
    > Dobin A, Davis CA, Schlesinger F, Drenkow J, Zaleski C, Jha S, Batut P, Chaisson M, Gingeras TR. STAR: ultrafast universal RNA-seq aligner. Bioinformatics. 2013 Jan 1;29(1):15-21. doi: 10.1093/bioinformatics/bts635. Epub 2012 Oct 25. PMID: 23104886; PMCID: PMC3530905.

* [Tabix](https://academic.oup.com/bioinformatics/article/27/5/718/262743)
    > Heng Li, Tabix: fast retrieval of sequence features from generic TAB-delimited files, Bioinformatics, Volume 27, Issue 5, 1 March 2011, Pages 718–719

## R packages

* [R](https://www.r-project.org/)
    >  R Core Team (2020). R: A language and environment for statistical computing. R Foundation for Statistical Computing, Vienna, Austria.

* [argparse](https://cran.r-project.org/package=argparse)
    >  Trevor L Davis (2022). argparse: Command Line Optional and Positional Argument Parser

* [Biostrings](https://bioconductor.org/packages/Biostrings)
    >  Pagès H, Aboyoun P, Gentleman R, DebRoy S. Biostrings: Efficient manipulation of biological strings

* [data.table](https://CRAN.R-project.org/package=data.table)
    >  Matt Dowle and Arun Srinivasan (2022). data.table: Extension of "data.frame".

* [deepSNV](https://pubmed.ncbi.nlm.nih.gov/22549840/)
    >  Gerstung M, Beisel C, Rechsteiner M, Wild P, Schraml P, Moch H, Beerenwinkel N. Reliable detection of subclonal single-nucleotide variants in tumour cell populations. Nat Commun. 2012 May 1;3:811. doi: 10.1038/ncomms1814. PMID: 22549840.

* [ggbeeswarm](https://CRAN.R-project.org/package=ggbeeswarm)
    >  Erik Clarke and Scott Sherrill-Mix (2017). ggbeeswarm: Categorical Scatter (Violin Point) Plots.

* [ggplot2](https://ggplot2.tidyverse.org)
    > H. Wickham. ggplot2: Elegant Graphics for Data Analysis. Springer-Verlag New York, 2016.

* [ggpubr](https://CRAN.R-project.org/package=ggpubr)
    >  Alboukadel Kassambara (2020). ggpubr: 'ggplot2' Based Publication Ready Plots.

* [Rsamtools](https://bioconductor.org/packages/Rsamtools)
    >  Martin Morgan, Hervé Pagès, Valerie Obenchain and Nathaniel Hayden (2020). Rsamtools: Binary alignment (BAM), FASTA, variant call (BCF), and tabix file import.

* [SeqinR](https://link.springer.com/chapter/10.1007/978-3-540-35306-5_10)
    >  Charif D, Lobry J (2007). “SeqinR 1.0-2: a contributed package to the R project for statistical computing devoted to biological sequences retrieval and analysis.” In Bastolla U, Porto M, Roman H, Vendruscolo M (eds.), Structural approaches to sequence evolution: Molecules, networks, populations, series Biological and Medical Physics, Biomedical Engineering, 207-232. Springer Verlag, New York. ISBN : 978-3-540-35305-8.
  
## Software packaging/containerisation tools

* [Singularity](https://pubmed.ncbi.nlm.nih.gov/28494014/)
    > Kurtzer GM, Sochat V, Bauer MW. Singularity: Scientific containers for mobility of compute. PLoS One. 2017 May 11;12(5):e0177459. doi: 10.1371/journal.pone.0177459. eCollection 2017. PubMed PMID: 28494014; PubMed Central PMCID: PMC5426675.
